src/compositor/shells/xdg_v6.rs

```diff
@@ -20,13 +20,24 @@ pub struct XdgV6ShellManager;
 impl XdgV6ShellManagerHandler for XdgV6ShellManager {
     fn new_surface(&mut self,
                    compositor: CompositorHandle,
-                   shell_surface: XdgV6ShellSurfaceHandle)
+                   mut shell_surface: XdgV6ShellSurfaceHandle)
                    -> Option<Box<XdgV6ShellHandler>> {
-        with_handles!([(compositor: {compositor})] => {
-            let server: &mut Server = compositor.into();
-            server.views
-                .push(View::new(Shell::XdgV6(shell_surface.into())));
+
+        let is_toplevel = with_handles!([(shell_surface: {&mut shell_surface})] => {
+            match shell_surface.state().unwrap() {
+                TopLevel(_) => true,
+                _ => false
+            }
         }).unwrap();
+
+        if is_toplevel {
+            with_handles!([(compositor: {compositor})] => {
+                let server: &mut Server = compositor.into();
+                server.views
+                    .push(View::new(Shell::XdgV6(shell_surface.into())));
+            }).unwrap();
+        }
+
         Some(Box::new(XdgV6::new()))
```

Comment:
That's a bug in wlroots-rs then, that should be fixed in a separate PR to wlroots-rs

Reply:
Interesting, if I try to modify the xdg example in wlroots-rs to return `None` it panics instead of segfaults. Are you sure it's segfaulting and not panicking?

Reply:
Actually now that I tried it, I couldn't get it to segfault. Instead it got caught in a loop throwing the panic for some reason instead of just terminating. Looking into it.

Reply:
Also for the record there should be a panic. We are adding a weak reference above in the `with_handles` and then immediantly dropping the backing storage by returning `None` here.  The shell surface only survives if we give it some callbacks to bind to.

Reply:
If you have to give an implementation, then maybe it should not be optional.

Reply:
Found the issue. We have a custom event loop hook in Way Cooler, so the mechanism I set up in wlroots-rs to catch the panic and deal with it properly doesn't work. I'll need to add a way in wlroots-rs to specify a shutdown mechanism. [See this wlroots-rs issue I made](https://github.com/swaywm/wlroots-rs/issues/157)

Reply:
> If you have to give an implementation, then maybe it should not be optional.

It's optional because that is also how you deny clients or devices or whatever. 

Let's say you want to have a super secure desktop or whatever. So you want to whitelist all input devices so nothing can act like an input device and act on your behalf. You want to deny any that don't fall under that. In wlroots-rs, you do that by returning `None` in those functions.